pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,136 @@
+
+package zfsmanager
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	libzfs "github.com/bicomsystems/go-libzfs"
+	"github.com/lf-edge/eve/api/go/info"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/vault"
+	"github.com/lf-edge/eve/pkg/pillar/zfs"
+)
+
+func getVdevNameStatus(dev libzfs.VDevTree, name string) (string, libzfs.VDevStat) {
+	if dev.Type != libzfs.VDevTypeDisk {
+		for _, d := range dev.Devices {
+			retDev, retStat := getVdevNameStatus(d, name)
+			if retDev != "" {
+				return retDev, retStat
+			}
+		}
+	} else {
+		// we can receive partition number
+		devName, err := types.GetDiskNameByPartName(dev.Name)
+		if err != nil {
+			log.Errorf("failed to get disk name by part name %s: %s", dev.Name, err)
+		}
+		if strings.Contains(devName, name) {
+			return devName, dev.Stat
+		}
+	}
+	return "", libzfs.VDevStat{}
+}
+
+func handleDiskConfigRestart(ctxArg interface{}, restartCounter int) {
+	log.Functionf("handleDiskConfigRestart(%d)", restartCounter)
+	ctx := ctxArg.(*zfsContext)
+	select {
+	case ctx.disksProcessingTrigger <- struct{}{}:
+	default:
+		log.Functionf("handleDiskConfigRestart: disksProcessingTrigger already triggered")
+	}
+}
+
+func processDisksTask(ctx *zfsContext) {
+
+	disksProcessingTicker := time.NewTicker(disksProcessingInterval)
+
+	wdName := agentName + "diskstask"
+	stillRunning := time.NewTicker(stillRunningInterval)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+
+	for {
+		select {
+		case <-disksProcessingTicker.C:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-ctx.disksProcessingTrigger:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func processDisks(ctx *zfsContext) error {
+	if vault.ReadPersistType() != types.PersistZFS {
+		return nil
+	}
+	disksConfig := ctx.subDiskConfig.GetAll()
+	if len(disksConfig) == 0 {
+		return nil
+	}
+	persistPool, err := libzfs.PoolOpen(vault.DefaultZpool)
+	if err != nil {
+		return fmt.Errorf("cannot open pool: %s", err)
+	}
+	defer persistPool.Close()
+	vdevTree, err := persistPool.VDevTree()
+	if err != nil {
+		return fmt.Errorf("cannot get vdev tree: %s", err)
+	}
+	for _, cfg := range disksConfig {
+		diskCfg := cfg.(types.EdgeNodeDiskConfig)
+		diskName := filepath.Base(diskCfg.Disk.Name)
+		if diskCfg.OldDisk != nil {
+			oldDiskName := filepath.Base(diskCfg.OldDisk.Name)
+			oldDevName, oldDevStat := getVdevNameStatus(vdevTree, oldDiskName)
+			if oldDevName != "" && diskCfg.Disk.Name != "" {
+				// if we found device oldDevName, replace it
+				log.Functionf("replacing %s with %s, old stat %s", oldDevName, diskCfg.Disk.Name, oldDevStat.State.String())
+				if stdout, err := zfs.ReplaceVDev(log, vault.DefaultZpool, oldDevName, diskCfg.Disk.Name); err != nil {
```

Comment:
Somewhere we need to check that there isn't a replace and associated resilver already running for this pool because we can't replace a second disk in a raidz1 before a previous replace has completed its resilver.

Reply:
So my question is whether we should make our code in zedmanager check such preconditions, or if we will just pass it to zfs and assume that zfs will generate errors for things which don't make sense (such as the above example of trying to replace multiple drives in a raidz1 at the same time).

If it fails the retry timer you have will try again I assume.

Reply:
Well, I think it still need additional testing on real workloads. Right now I expect that it will be handled by retry. Added comments about that assumption

Reply:
ok